
# <span style='color:midnightblue'> Python para Actuarios Vida   </span>

![ADRISK](Imagenes/KN.png)


## <span style='color:midnightblue'> Sesión 02 : Matemática Financiera  </span>

Se desarrolla los conceptos básicos de los vectores de flujo de efectivo y su valoración con factores de descuento. Luego, evaluará los flujos en función de su valor actual neto y creará su propia calculadora financiera. Finalmente, aprenderá acerca de las tasas de interés fijas y variables; y equivalencia de tasas en diferentes periodos.

### 1. Flujo de efectivo

Fije una unidad de capital y una unidad de tiempo:   
- **0** es el momento presente;
- ***k*** es *k* unidades de tiempo en el futuro (por ejemplo, años, meses, trimestres). 

Cantidad de dinero recibido o pagado en el momento ***k***:  


- $C_{k}$
- el flujo de caja en el momento ***k***

En general: para un vector de flujo de caja ($C_{0}$, $C_{1}$, $C_{2}$, $C_{3}$, $C_{4}$, ... , $C_{N-1}$, $C_{N}$):



<table style="width:40%">
<tr>
<th> Periodo </th>
<th> 0 </th>
<th> 1 </th>
<th> 2 </th>
<th> 3 </th>
<th> 4 </th>
<th> ... </th>      
<th> N-1 </th>
<th> N </th>     
</tr>

<tr>
<td> Flujo </td>
<td> $C_{0}$ </td>    
<td> $C_{1}$ </td>    
<td> $C_{2}$ </td>    
<td> $C_{3}$ </td>
<td> $C_{4}$ </td>    
<td> ... </td>  
<td>  $C_{ N-1 }$  </td>      
<td>  $C_{ N }$  </td>
</tr>
</table>    


####  Ejercicio 01 
Defina el siguiente flujo :

| Periodo | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| Flujo | 500 | 400 | 300 | 200 | 200 | 200 | 200 | 200 |

In [1]:
#importamos libreria de operaciones con números
import numpy as np                # permite hacer operaciones mediante funciones númericas
import pandas as pd               # ofrece estructuras de datos y operaciones para manipular tablas numéricas y series de tiempo


In [4]:
# opción 1 : creación de matriz
flujo_m = 200*np.ones(8)
print(flujo_m)

# modificar valores
flujo_m[0] = 500
flujo_m[1] = 400
flujo_m[2] = 300
print(flujo_m)


[200. 200. 200. 200. 200. 200. 200. 200.]
[500. 400. 300. 200. 200. 200. 200. 200.]


In [5]:
# opción1.1 : lo convertimos la matriz en serie
flujo_s1 = pd.Series(flujo_m)
print(flujo_s1)

0    500.0
1    400.0
2    300.0
3    200.0
4    200.0
5    200.0
6    200.0
7    200.0
dtype: float64


In [6]:
# opción2: creamos una serie inicial y cambiamos montos
flujo_s2 = pd.Series([200]*8)
flujo_s2
# modificar valores
flujo_s2[0] = 500
flujo_s2[1] = 400
flujo_s2[2] = 300
print(flujo_s2)

0    500
1    400
2    300
3    200
4    200
5    200
6    200
7    200
dtype: int64


In [7]:
# opción3 : creamos el flujo de forma directa (serie)
flujo_s3 = pd.Series([500] + [400]+ [300] + [200]*5)
print(flujo_s3)

0    500
1    400
2    300
3    200
4    200
5    200
6    200
7    200
dtype: int64


In [15]:
# opción4 : creamos el flujo de forma directa (lista)
flujos_l = [500,400,300] + [200]*5

print(flujos_l)

[500, 400, 300, 200, 200, 200, 200, 200]


####  Ejercicio 02
Defina el siguiente flujo :

| Periodo | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| Flujo | 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 |

In [8]:
#  creamos el flujo de forma directa (serie)

flujo_e2 = pd.Series([100]*8)
print(flujo_e2)

for x in range(8):
    flujo_e2[x] = 100 + x*100
    
print(flujo_e2)    

0    100
1    100
2    100
3    100
4    100
5    100
6    100
7    100
dtype: int64
0    100
1    200
2    300
3    400
4    500
5    600
6    700
7    800
dtype: int64


### 2. Tasa de interés
Hechos cruciales:
- **¡El momento del flujo de caja importa!** ---    ---  ¡timing of cash flow matter!
- **¡El valor temporal del dinero importa!** ---    ---   ¡time value of money matter!  

La **tasa de interés** determina el crecimiento del dinero.

#### Acumulación
**i** es la tasa de interés constante.

 | 0 |  | 1 | 
| :---: | :---:  | :---: | 
 |100 PEN |$\Rightarrow$ |  **¿x?**  |

In [9]:
i = 0.03
x = 100*(1+i)
print(x)

103.0


#### Descuento
El factor de descuento es $v = \frac{1}{1+i} $ .

| 0 |  | 1 |
| :---: | :---:  | :---: | 
| **¿y?** |$\Leftarrow$ | 100 PEN |

In [10]:
v = 1/(1+i)
y = 100*v
print(y)

97.0873786407767


### De un período de tiempo a ***k*** períodos de tiempo
-  #### Acumulación

      - El valor en el  ***momento k*** de **100 PEN** pagado en el ***momento 0***
   
      - $= (1+i)^K = v^{-K} $
      
| 0 |  | ***k*** | 
| :---: | :---:  | :---: | 
 |100 PEN |$\Rightarrow$ |  **¿x?**  |

-  #### Descuento

      - El valor en el  ***momento 0*** de **100 PEN** pagado en el ***momento k***
   
      - $= (1+i)^{-K} = v^{K} $
 
 
| 0 |  | ***k*** |
| :---: | :---:  | :---: | 
| **¿y?** |$\Leftarrow$ | 100 PEN |


In [11]:
# Acumulación
k=2
x1 =100*(1+i)**k
x2 = 100*v**-k

print(x1)
print(x2)

106.08999999999999
106.08999999999999


In [12]:
# Descuento
k=2
y1 =100*(1+i)**-k
y2 = 100*v**k

print(y1)
print(y2)

94.25959091337543
94.25959091337543


### 3. El valor presente de un vector de flujo de efectivo
¿Cuál es el valor en k = 0 del  flujo de caja definido en el ejercicio 1?

In [13]:
# en el ejercicio 1 el vector se creo con cualquiera de los nombres : flujo_m  flujo_s1 flujo_s2 flujo_s3 flujo_l
print(flujo_s3)
print("cantidad de periodos")

per = len(flujo_s3)
print(per)

print("-----------")

print("necesita definir el vector de exponentes")
exp = pd.Series(range(per))  #exp = pd.Series(range(0,8,1))   # exp = pd.Series(range(8))
print(exp)

print("-----------")


print("flujos en valor presente")

va_pre = flujo_s3*1/(1+i)**exp
print(va_pre)

print("valor presente")
sum(va_pre)




0    500
1    400
2    300
3    200
4    200
5    200
6    200
7    200
dtype: int64
cantidad de periodos
8
-----------
necesita definir el vector de exponentes
0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    7
dtype: int64
-----------
flujos en valor presente
0    500.000000
1    388.349515
2    282.778773
3    183.028332
4    177.697410
5    172.521757
6    167.496851
7    162.618302
dtype: float64
valor presente


2034.4909392392376

Flujo de efectivo

| Periodo | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| Flujo | 500 | 400 | 300 | 200 | 200 | 200 | 200 | 200 |
| **Descuento**|$v^0$  |$v^1$  |$v^2$ |$v^3$ |$v^4$  | $v^5$ | $v^6$ | $v^7$ |
| **En el momento 0** | 500 | 388.34 | 282.77 | 183.02 | 177.69 | 172.52 | 167.49 | 162.61 |
| **Valor Presente**| 2034.49 |  | | |  |  |  |  |


In [14]:
# opción 2 : como Dataframe

cash_flow2 = pd.DataFrame({"Flujo":flujo_s3}) # creamos una columna llamada flujo con los valores de la serie

cash_flow2["year"] = range(len(flujo_s3))

cash_flow2["discont fac"] = 1/(1+i)**cash_flow2["year"] # or cash_flow2["discontfac"] = 1/(1+i)**cash_flow2.index

cash_flow2["flujo descon"] = cash_flow2["Flujo"]*cash_flow2["discont fac"]

sum(cash_flow2["flujo descon"])
print(cash_flow2)

   Flujo  year  discont fac  flujo descon
0    500     0     1.000000    500.000000
1    400     1     0.970874    388.349515
2    300     2     0.942596    282.778773
3    200     3     0.915142    183.028332
4    200     4     0.888487    177.697410
5    200     5     0.862609    172.521757
6    200     6     0.837484    167.496851
7    200     7     0.813092    162.618302


### 4. Valuación: ¿acumulación o descuento?
Denote: $v(s, t)$ el valor en el **momento s** de **1 PEN** pagado en el **momento t**.   

   - $s<t$: un factor de descuento
| $s$ |  | $t$ |
| :---: | :---:  | :---: | 
| **¿y?** |$\Leftarrow$ | 1 PEN |
   - $s>t$: un factor de acumulación
| $t$ |  | $s$ | 
| :---: | :---:  | :---: | 
 |100 PEN |$\Rightarrow$ |  **¿x?**  |   

**Determine**: $v (2, 4)$   
valor en el **momento 2** de **1 PEN** pagado en el **momento 4**

In [15]:
i = 0.03
v = 1 / ( 1 + i)

s = 2 
t = 4

res1 = v**(t-s)
res2 = (1 + i) **-(t-s)

print(res1)
print(res2)

0.9425959091337544
0.9425959091337544


**Determine**:  𝑣(6,3)    
valor en el **momento 6** de **1 PEN** pagado en el **momento 3**

In [16]:
s = 6 
t = 3
 
res3 = v**(t-s)
res4 = (1 + i) **-(t-s)

print(res3)
print(res4)

1.092727
1.092727


### 5. Valoración de un vector de flujo de efectivo

<table style="width:40%">
<tr>
<th> Periodo </th>
<th> 0 </th>
<th> 1 </th>
<th> 2 </th>
<th> 3 </th>
<th> 4 </th>
<th> ... </th>      
<th> N-1 </th>
<th> N </th>     
</tr>

<tr>
<td> Flujo </td>
<td> $C_{0}$ </td>    
<td> $C_{1}$ </td>    
<td> $C_{2}$ </td>    
<td> $C_{3}$ </td>
<td> $C_{4}$ </td>    
<td> ... </td>  
<td>  $C_{ N-1 }$  </td>      
<td>  $C_{ N }$  </td>
</tr>
</table>    

- El valor en el periodo $n$ :

     $\sum \limits _{k=0} ^{N} C_{k}.v(n,k) $
     
     con $0 ≤ n ≤ N$.
     
     
     
- Valor presente $(n=0)$ y valor acumulado $(n=N)$.     

####  Ejercicio 03 
Defina el valor del siguiente flujo en el momento 03 :

| Periodo | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| Flujo | 500 | 0 | 300 | 0 | 0 | 0 | 0 | 200 |

In [17]:
# como funcion / creación de funciones
def discount(s,t,i):
     return (1+i)**(s-t) #  (1+i)**-(t-s) =  v**(t-s) // v = (1+i)**-1
 
    
print(discount(2,4,i=0.03) ) 
print(discount(6,3,i=0.03) )  

0.9425959091337544
1.092727


In [18]:
# OPCION 01: MEDIANTE LA SUMA DE VALORES
## valor en periodo 3 de 500 en 0 , 300 en 2, y 200 en 7
momento = 3

valor3 = 500*discount(momento,0,0.03)+ 300*discount(momento,2,0.03) + 200*discount(momento,7,0.03)
valor3

1033.0609095831378

In [19]:
# OPCION 02: MEDIANTE LA CREACION DE UNA SERIE
momento =3
valor3s = pd.Series([500,0,300,0,0,0,0,200])* discount(momento,pd.Series(range(0,8,1)), 0.03)
sum(valor3s)

1033.0609095831378

In [20]:
# OPCION 03: MEDIANTE DATAFRAME
momento =3

flujo_e3 =pd.DataFrame({"Flujo":[500,300,200],"year":[0,2,7]})

print(flujo_e3)

flujo_e3["valor per 3"] = flujo_e3["Flujo"]*discount(momento,flujo_e3["year"],0.03)
res_e3 = sum(flujo_e3["valor per 3"])

print(flujo_e3)
print(res_e3)


print('.......................')
print('Filtrar filas')

print(flujo_e3.loc[0])
print('..........')

print(flujo_e3[flujo_e3["year"]== 0])

   Flujo  year
0    500     0
1    300     2
2    200     7
   Flujo  year  valor per 3
0    500     0    546.36350
1    300     2    309.00000
2    200     7    177.69741
1033.0609095831378
.......................
Filtrar filas
Flujo          500.0000
year             0.0000
valor per 3    546.3635
Name: 0, dtype: float64
..........
   Flujo  year  valor per 3
0    500     0     546.3635


In [21]:
# OPCION 03.1: MEDIANTE DATAFRAME - convertimos el año en indice

momento =3

flujo_e3a =pd.DataFrame({"Flujo":[500,300,200]}, index = [0,2,7])
print(flujo_e3a)

flujo_e3a["valor per 3"] = flujo_e3a["Flujo"]*discount(momento,flujo_e3a.index,0.03)
res_e3a = sum(flujo_e3["valor per 3"])

print(flujo_e3a)
print(res_e3a)


flujo_e3a.loc[7]   # al colocar el periodo como index me facilita el filtrado

   Flujo
0    500
2    300
7    200
   Flujo  valor per 3
0    500    546.36350
2    300    309.00000
7    200    177.69741
1033.0609095831378


Flujo          200.00000
valor per 3    177.69741
Name: 7, dtype: float64

### 5. Cambios en la  estructura de períodos y tasas

Dos preguntas:
1. ¿Cómo lidiar con las tasas de interés cuando se aplica un cambio de período (por ejemplo, de años a meses)?
2. ¿Cómo pasar de una tasa de interés constante a una tasa que cambia con el tiempo?

**De tipos de interés anual a mensual**   

   ¿Cómo derivar  la tasa  $i^*_{m}$ aplicable al período de un mes del año?
   ![m](imagenes/m-y.png)
   Entonces :
   
   -    $1 + i = (1 + i^*_{m})^m ⇔ i = (1 + i)^{1/m} − 1.$

In [22]:
# interés anual 
i = 0.03
# calcule interés mensual
m_i = (1 + i)**(1/12)-1
print(m_i)


0.0024662697723036864


In [23]:
# De interés mensual a anual
(1+m_i)**12 -1 

0.030000000000001137

**Tasas de interés no constantes**

Observaciones:
- las tasas de interés no son necesariamente constantes;
- la estructura temporal de las tasas de interés o la curva de rendimiento.   

¡Incorporemos esto en nuestra notación!
![n-i](imagenes/n_i.png)

**Ejercicio 04**   

Defina el siguiente vector de tasas:
![n-i](imagenes/e4.png)

In [24]:
interest = pd.Series([0,0.04,0.03,0.02,0.01])
interest

0    0.00
1    0.04
2    0.03
3    0.02
4    0.01
dtype: float64

In [50]:
# Defina el vector que contiene el inverso de 1 más la tasa de interés
yearly_discount_factors=(1+interest)**-1
yearly_discount_factors


0    1.000000
1    0.961538
2    0.970874
3    0.980392
4    0.990099
dtype: float64

In [51]:
# Defina los factores de descuento al tiempo 0 usando cumprod()
discount_factors = np.cumprod(yearly_discount_factors)
discount_factors

0    1.000000
1    0.961538
2    0.933532
3    0.915228
4    0.906166
dtype: float64

### 6. Funciones / Librerias financieras

**6.1 Cashflows**

Las funciones de este módulo se utilizan para ciertos cálculos de interés compuesto para un flujo de efectivo bajo las siguientes restricciones:

- Los períodos de pago coinciden con los períodos de capitalización.  
- Los pagos se realizan a intervalos regulares.   
- Los pagos son una cantidad constante.   
- La tasa de interés es la misma durante todo el período de análisis

In [25]:
import cashflows

**pvfv**: 
calcula el valor faltante en la ecuación $fval = pval*(1 + nrate)^{nper}$

In [27]:
print(cashflows.pvfv(pval=100,nper=1,nrate=10))
print(cashflows.pvfv(fval=110,nper=1,nrate=10))

-110.00000010000001
-100.00000009090908


**pvpmt**: calcula el valor faltante $(pmt, pval, nper, nrate)$ en un modelo que relaciona un valor presente y una secuencia finita de pagos realizados al final del período (pagos atrasados o anualidades ordinarias)

In [28]:
z = cashflows.pvpmt(𝑝𝑣𝑎𝑙=100,𝑛𝑝𝑒𝑟=5,𝑛𝑟𝑎𝑡𝑒=3)
print(z)

a = cashflows.pvpmt(pmt=z,𝑛𝑝𝑒𝑟=5,𝑛𝑟𝑎𝑡𝑒=3)
print(a)

-21.835457140057592
100.0


**amortize**  calcula y devuelve las columnas de un programa de amortización de un préstamo.  
La función devuelve el pago de intereses, el reembolso del principal, el pago periódico y el saldo al final de cada período.

In [29]:
x = cashflows.amortize(pval=100, nrate=10, nper=5, fval=0) 
print(x)


   Beg_Balance  Payment  Interest  Principal  Final_Balance
0       100.00     0.00      0.00       0.00         100.00
1       100.00   -26.38     10.00     -16.38          83.62
2        83.62   -26.38      8.36     -18.02          65.60
3        65.60   -26.38      6.56     -19.82          45.78
4        45.78   -26.38      4.58     -21.80          23.98
5        23.98   -26.38      2.40     -23.98           0.00


In [29]:
print(sum(x['Principal']))

-100.0


**6.2 Numpy financial**

El paquete numpy-financial es una colección de funciones financieras elementales:   
 -    **fv**(rate, nper, pmt, pv[, when])  $\Rightarrow$        $\Rightarrow$    Determina el valor futuro  
 - 
 - **ipmt**(rate, per, nper, pv[, fv, when]) $\Rightarrow$  Determina los intereses de un pago
 - 
 - **irr**(values) $\Rightarrow$  Devuelve la Tasa Interna de Retorno (TIR).   
 - 
 - **mirr**(values, finance_rate, reinvest_rate)   $\Rightarrow$ Devuelve la Tasa Interna de Retorno Modificada   
 - 
 - **nper**(rate, pmt, pv[, fv, when])  $\Rightarrow$ Calcula la cantidad de pagos periódicos.  
 - 
 - **npv**(rate, values) $\Rightarrow$ Devuelve el VAN (valor actual neto) de una serie de flujos de efectivo 
 - 
 - **pmt**(rate, nper, pv[, fv, when]) $\Rightarrow$ Calcula el pago contra el capital del préstamo más los intereses  
 - 
 - **ppmt**(rate, per, nper, pv[, fv, when]) $\Rightarrow$ Calcula el pago contra el capital del préstamo.   
 - 
 - **pv**(rate, nper, pmt[, fv, when]) $\Rightarrow$ Calcule el valor presente.  
 - 
 - **rate**(nper, pmt, pv, fv[, when, guess, tol, …]) $\Rightarrow$ Calcule la tasa de interés por período  


In [31]:
import numpy_financial as npf

In [33]:
interest_rate = 0.05/12
n_periods = 10*12
payment_per_month = -200
present_value = -200

In [34]:
future_value = npf.fv(interest_rate, n_periods, payment_per_month, present_value)
print(round(future_value))

31386


In [36]:
# importa el orden de los argumento, se recomienda usar la asignaciones
future_value2 = npf.fv(pv = 200,pmt=200,nper=120,rate=0.05/12)
print(round(future_value2))

-31386


#### <span style='color:midnightblue'>  Edwin Fernandez - ADRISK
<span style='color:midnightblue'> Programa de Especialización Ejecutiva - Python para Actuarios Vida  </span>
***